python/ray/tune/function_runner.py

```diff
@@ -229,6 +232,24 @@ def get_checkpoint(self):
     def _start(self):
         self._last_report_time = time.time()
 
+    def report(self, metrics: Dict, checkpoint: Optional[Checkpoint] = None) -> None:
+        # TODO(xwjiang): Tons of optimizations.
+        if not self._iter:
+            self._iter = 0
+        if checkpoint:
+            checkpoint_dir = self.make_checkpoint_dir(step=self._iter)
+            self.set_checkpoint(checkpoint_dir)
+            checkpoint.to_directory(checkpoint_dir)
```

Comment:
We may have to drop a new marker here (`to_directory()` may remove the existing dir), i.e.

```suggestion
            checkpoint.to_directory(checkpoint_dir)
            open(os.path.join(checkpoint_dir, ".is_checkpoint"), "a").close()
```

Reply:
(note I'm refactoring this right now so I can make this a bit nicer once that's done)